In [144]:
import webbrowser, bs4, requests
import re
import pandas as pd
import numpy as np

In [5]:
# Creating empty lists in-order to append data scrapped from url's.
# The number of lists depends on the number of features you are extracting from the url.

title=[]
location=[]
price=[]
price_per_sqft=[]
area_in_sqft=[]
building_status=[]

In [6]:

n_pages = 0
for page in range(3,2016):  #The no. of pages you wanted to scrap considering pages available in the url.
    n_pages += 1
    url = 'http://www.makaan.com/hyderabad-residential-property/buy-property-in-hyderabad-city?budget=,'+'&page='+str(page)
    # This for gets us into next page after every iteration.
    r = requests.get(url) # Here r is response data.
    page_html = bs4.BeautifulSoup(r.text, 'html.parser') 
    house_containers = page_html.find_all("li",class_="cardholder") # Class containing overall data of a property.
    for data in house_containers:
################################   TITLE   ########################################  

        type_=data.find_all("a",class_="typelink") # Class containing title name of a property.
        for i in type_:
            d=i.text
            title.append(d) # Appending the extracted data into a list.
            
################################ LOCATION  #######################################

        location_=data.find_all("a",class_="loclink")
        for i in location_:
            d=i.text
            r= d.split(',',)[0] # Splitting the obtained text and returning the first element of text.
            location.append(r)
            
################################    PRICE    ######################################

        cost=data.find_all("td",class_="price")
        for i in cost:
            d=i.text
            if 'L' in d:
                a = d.split()[0]
                price.append(float(a)) # Type conversion of obtained data.
            elif 'Cr' in d:
                b = d.split()[0]
                price.append(float(b) * 100) # Coverting CRORES into LAKHS.
            else:  # If no value found in the class return value 0.
                e=0
                price.append(float(e)) 
                
############################   PRICE PER SQFT   ###################################

        rate_sqft=data.find_all('td',class_="lbl rate")
        for i in rate_sqft:
            d=i.text
            res= d.split('/',)[0]
            r=re.sub(",","" ,res) # Substitute no character in the place of ,(comma).
            price_per_sqft.append(r)
            
############################   AREA IN SQFT    ####################################

        area_sqft=data.find_all('td',class_="size")
        for i in area_sqft:
            d=i.text
            area_in_sqft.append(d)
            
##########################   BUILDING STATUS    ###################################

        building_status_=data.find_all("td",class_="val")
        for i in building_status_:
            d=i.text
            building_status.append(d)

###################################################################################

print('You scraped {} pages containing {} properties.'.format(n_pages, len(title)))
# Returns the number of pages scrapped and time taken for scarpping those pages.


You scraped 2013 pages containing 4840 properties.


In [ ]:
n_pages = 0
agent_ratings=[]
for page in range(3,2016):
    n_pages += 1
    url = 'http://www.makaan.com/hyderabad-residential-property/buy-property-in-hyderabad-city?budget=,'+'&page='+str(page)
    r = requests.get(url)
    if (r.status_code == 200):
        page_html = bs4.BeautifulSoup(r.text, 'html.parser')
        house_containers = page_html.find_all("li",class_="cardholder")
        for i,data in enumerate(house_containers):
            agent_rating_ = data.find_all('div', class_="rating")
            if len(agent_rating_)==1:
                for j in agent_rating_:
                    d=j.text
                    agent_ratings.append(d)
            else:
                 agent_ratings.append(int('0'))  
    else:
        print("error in response")



In [15]:
# Defining a dataframe.

df=pd.DataFrame()

In [23]:
# Returns data in form a dataframe with columns of specified names containing the assigned values from the list.

df['title'] = title 
df['location'] = location
df['price(L)'] = price
df['rate_persqft'] = price_per_sqft
df['area_insqft'] = area_in_sqft
df['building_status'] = building_status
df['agent_rating'] = agent_ratings

ValueError: Length of values does not match length of index

In [42]:

# Returns data in form a dataframe with columns of specified names containing the assigned values from the list.

df['title']=title 
df['location']=location
df['price_lakhs']=price
df['rate_persqft']=price_per_sqft
df['area_insqft']=area_in_sqft
df['building_status']=building_status


In [2]:
n_pages = 0
agent_ratings=[]
for page in [1]:
    n_pages += 1
    url = 'http://www.makaan.com/hyderabad-residential-property/buy-property-in-hyderabad-city?budget=,'+'&page='+str(page)
    r = requests.get(url)
    if (r.status_code == 200):
        page_html = bs4.BeautifulSoup(r.text, 'html.parser')
        house_containers = page_html.find_all("li",class_="cardholder")
        for i,data in enumerate(house_containers):
            agent_rating_ = data.find_all('div', class_="rating")
            if len(agent_rating_)==1:
                for j in agent_rating_:
                    d=j.text
                    agent_ratings.append(d)
            else:
                 agent_ratings.append(int('0'))  
    else:
        print("error in response")


In [3]:
d=df.to_csv('projectfinal.csv')

NameError: name 'df' is not defined

# loading the data frame

In [129]:
df=pd.read_csv("projectfinal1.csv", index_col='Unnamed: 0')
df

title     location  price(L)  rate_persqft  \
0             Residential Plot  Maheshwaram     45.00          1111   
1             Residential Plot     Kondakal     34.00          2361   
2             Residential Plot    Bibinagar     14.19           944   
3             Residential Plot   Gachibowli    330.00          3333   
4             Residential Plot    Shadnagar      6.72           466   
...                        ...          ...       ...           ...   
40255          2 BHK Apartment  Chandanagar     47.00          4563   
40256  2 BHK Independent House  Kundanpally     25.74          3961   
40257          3 BHK Apartment  Peeramcheru     81.00          4500   
40258          2 BHK Apartment     Nizampet     34.50          3833   
40259          2 BHK Apartment  Chandanagar     40.00          4444   

       area_insqft     building_status  agent_rating  
0             4050                 New           4.6  
1             1440                 New           3.7  
2             1503                 New           4.9  
3             9900                 New           0.0  
4             1440                 New           4.7  
...            ...                 ...           ...  
40255         1030       Ready to move           0.0  
40256          650       Ready to move           0.0  
40257         1800       Ready to move           0.0  
40258          900  Under Construction           0.0  
40259          900  Under Construction           0.0  

[40260 rows x 7 columns]

In [130]:
df.shape
unclean_df = df.copy()
unclean_df.describe()

price(L)   rate_persqft    area_insqft  agent_rating
count  40260.000000   40260.000000   40260.000000  40260.000000
mean      65.331670    3297.212072    2360.068728      3.574724
std      126.247236    3850.078324    4190.331350      1.953369
min        0.000000     100.000000     100.000000      0.000000
25%       20.000000    1018.000000    1290.000000      3.400000
50%       37.170000    2222.000000    1674.000000      4.600000
75%       78.630000    4800.000000    2250.000000      4.900000
max     7000.000000  172541.000000  215278.000000      5.000000

**Observations from unclean dataset:**

price(L):

* mean of prices is 65.33 L
* maximum price is 7000 L i.e, 70 cr.
* minimum price is 0
* 25 percentile is 20 L i.e, 25% of prices lie below 20 L.
* 50 percentile is 37.17 L i.e, 50% of prices lie below 37.17 L.
* 75 percentile is 78.63 L i.e, 75% of prices lie below 78.63 L.
* standard deviation(std) is 126.24

rate_persqft:

* mean ratepersqft is 3297.21/-
* maximum ratepersqft is 172541/-
* minimum ratepersqft is 100/-
* 25 percentile is 1018 i.e, 25% of ratepersqft's lie below 1018/-
* 50 percentile is 2222 i.e, 50% of ratepersqft's lie below 2222/-
* 75 percentile is 4800 i.e, 75% of ratepersqft's lie below 4800/-
* standard deviation(std) is 3850.07

area_insqft:

* mean area is 2360.06 sqft
* maximum area is 215278 sqft.
* minimum area is 100 sqft.
* 25 percentile is 1290 sqft i.e, 25% of area's lie below 1290 sqft.
* 50 percentile is 1674 sqft i.e, 50% of area's lie below 1674 sqft.
* 75 percentile is 2250 sqft i.e, 75% of area's lie below 2250 sqft.
* standard deviation(std) is 4190.33

agent_rating:

* mean of rating is 3.5
* maximum rating is 5.0
* minimum rating is 0.0
* 25 percentile is 3.4 i.e, 25% of rating's lie below 3.4
* 50 percentile is 4.6 i.e, 50% of rating's lie below 4.6
* 75 percentile is 4.9 i.e, 75% of rating's lie below 4.9
* standard deviation(std) is 1.95

In [131]:
df.dtypes

title               object
location            object
price(L)           float64
rate_persqft         int64
area_insqft          int64
building_status     object
agent_rating       float64
dtype: object

observations: 
1. **title** : Defines the title name of the property. datatype is string
2. **location** : Defines about the locality(area) of the property. datatype is string
3. **price(L)** : Defines the price of a property in lakhs. datatype is float.
4. **rate_persqft** : Defines the rate per unit sqft of property. datatype is int.
5. **area_insqft** : Defines the total area of property. datatype is int.
6. **building_status** : Define about the current status of builiding. datatype is string
7. **agent_rating** : Defines about the rating of agent given by the customers. datatype is float.

# 2. Cleaning Data 

In [132]:
# drop the duplicate entries in dataset
df = df.drop_duplicates()
# as the index will be changed after dropping rows
df = df.reset_index()


In [133]:
# droping index column created due to reset_index()
df = df.drop(columns=['index'])

In [134]:
df.shape

(25715, 7)

observations:
* Before deleting duplicate entries total datapoints were 40260
* After deleting duplicate entries total datapoints reduced to 25715
* That is we have 14545 (40260-25715) duplicate entries.

In [135]:
# checking for null values
df.isnull().sum()

title              0
location           0
price(L)           0
rate_persqft       0
area_insqft        0
building_status    0
agent_rating       0
dtype: int64

No Null values

In [136]:
df.describe()

price(L)   rate_persqft    area_insqft  agent_rating
count  25715.000000   25715.000000   25715.000000  25715.000000
mean      68.963354    3491.056621    2328.315497      3.363387
std      140.980749    4218.740006    4836.964889      2.017042
min        0.000000     100.000000     100.000000      0.000000
25%       20.000000    1111.000000    1250.000000      0.000000
50%       38.980000    2556.000000    1645.000000      4.500000
75%       76.000000    4962.000000    2250.000000      4.800000
max     7000.000000  172541.000000  215278.000000      5.000000

from above we can see that minimum value is 0 which is not possible. Lets investigate

In [137]:
index_list = df[df['price(L)']==0].index.tolist() 
index_list

[7860, 9992, 13947, 14320]

lets calculate Price for the rows from given 'rate_persqft' and 'area_insqft' values

In [138]:
price1=[] 
c=0
for i in (df['price(L)']):
    if i == 0:
        a=df.loc[index_list[c],'rate_persqft'] # Returns the value of ratepersqft at nth location.
        b=df.loc[index_list[c],'area_insqft'] # Returns value of area at nth location.
        m=np.round(float((a*b)/100000),2) # multiplies area and ratepersqft and coverts type of output to float.
        df.loc[index_list[c], 'price(L)'] = m # Appends the obtained value from mulitplication to nth location.
        c+=1
        price1.append(m)

In [139]:
price1

[0.45, 0.5, 0.32, 0.16]

In [140]:
df[df['price(L)']==0]

Empty DataFrame
Columns: [title, location, price(L), rate_persqft, area_insqft, building_status, agent_rating]
Index: []

In [141]:
df.iloc[index_list,:]

title                location  price(L)  rate_persqft  \
7860   Residential Plot         Vanasthalipuram      0.45           100   
9992   Residential Plot           Jubilee Hills      0.50           333   
13947  Residential Plot              Kukatpally      0.32           213   
14320  Residential Plot  Nagole Bandlaguda Road      0.16           106   

       area_insqft building_status  agent_rating  
7860           450             New           4.4  
9992           150             New           2.9  
13947          150             New           2.4  
14320          150             New           2.0

Hence all the value for Price(L) = 0 are replaced with correct values

In [142]:
#checking whether there's any change in mean and other values after data cleaning.

print("\t\t\t Before Cleaning")
print(unclean_df.describe())
print('-'*63)
print('*'*63)
print('-'*63)
print("\t\t\t After Cleaning")
print(df.describe())

			 Before Cleaning
           price(L)   rate_persqft    area_insqft  agent_rating
count  40260.000000   40260.000000   40260.000000  40260.000000
mean      65.331670    3297.212072    2360.068728      3.574724
std      126.247236    3850.078324    4190.331350      1.953369
min        0.000000     100.000000     100.000000      0.000000
25%       20.000000    1018.000000    1290.000000      3.400000
50%       37.170000    2222.000000    1674.000000      4.600000
75%       78.630000    4800.000000    2250.000000      4.900000
max     7000.000000  172541.000000  215278.000000      5.000000
---------------------------------------------------------------
***************************************************************
---------------------------------------------------------------
			 After Cleaning
           price(L)   rate_persqft    area_insqft  agent_rating
count  25715.000000   25715.000000   25715.000000  25715.000000
mean      68.963410    3491.056621    2328.315497      3.363387
s

**Difference b/w Observations:**

price(L):

* mean of prices changed from 65.33 to 68.96 L.
* minimum price from 0 L to 0.16 L.
* 25 percentile stays same 20 L i.e, 25% of prices lie below 20 L.
* 50 percentile changed from 37.17 to 38.98 L i.e, 50% of prices lie below 38.98 L.
* 75 percentile changed from 78.63 to 76.60 L i.e, 50% of prices lie below 76.60 L.
* standard deviation(std) raised to 140.98 from 126.24

rate_persqft:

* mean ratepersqft from 3297.21 to 3491/-
* 25 percentile from 1018 to 1111/-
* 50 percentile from 2222 to 2556/-
* 75 percentile from 4800 to 4962/-
* standard deviation(std) from 3850.07 to 4218.05

area_insqft:

* mean area from 2360.06 to 2328.31
* 25 percentile from 1290 sqft to 1250 sqft.
* 50 percentile from 1674 sqft to 1645 sqft.
* 75 percentile stays same i.e, 2250 sqft.
* standard deviation(std) from 4190.33 to 4836.96

agent_rating:

* mean of rating from 3.5 to 3.3
* 25 percentile from 3.4 to 0.0
* 50 percentile from 4.6 to 4.5
* 75 percentile from 4.9 to 4.8
* standard deviation(std) from 1.95 to 2.01

In [143]:
# saving the Clean dataset
df.to_csv('Clean_data.csv')